invenio_vocabularies/services/facets.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
+#
+# Invenio-Vocabularies is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Vocabulary facets."""
+
+from flask_principal import AnonymousIdentity
+from invenio_i18n.ext import current_i18n
+from marshmallow_utils.fields.babel import gettext_from_dict
+from speaklater import make_lazy_string
+
+from ..proxies import current_service
+
+
+def lazy_get_label(vocab_item):
+    """Lazy evaluation of a localized vocabulary label."""
+    params = {
+        "locale": current_i18n.locale,
+        "default_locale": "en"
+    }
+
+    return make_lazy_string(
+        gettext_from_dict, vocab_item, **params)
+
+
+class VocabularyLabels:
+    """Fetching of vocabulary labels for facets."""
+
+    def __init__(self, vocabulary, cache=False):
+        """Initialize the labels."""
+        self.vocabulary = vocabulary
+        self.cache = cache
+        self.fields = ["id", "title"]  # not configurable
+
+    def __call__(self, ids):
+        """Return the mapping when evaluated."""
+        identity = AnonymousIdentity()
+
+        if not self.cache:
+            vocabs = current_service.read_many(
+                identity, ids, fields=self.fields)
+        else:
+            vocabs = current_service.read_all(identity, fields=self.fields)
+
+        labels = {}
+        vocab_list = list(vocabs.hits)  # the service returns a generator
+        ids = set(ids)
+        seen = set()
+        for vocab in vocab_list:
+            # cannot loop over ids because vocab id is inside each item
+            if len(ids) == len(seen):
+                break
+
+            id_ = vocab["id"]
+            if id_ in ids:
+                labels[id_] = lazy_get_label(vocab["title"])
+                seen.add(id_)
+
+        return labels
```

Comment:
This seemed complicated, so hopefully I didn't *oversimplify*. 

```suggestion
        return {
            vocab["id"]: lazy_get_label(vocab["title"])
            for vocab in vocab.hits if vocab["id"] in set(ids)
        }
```

Previous code seemed to imply there could be more than 1 result with same "id" in `vocab.hits`:
- two vocab hits with same "id" have same other values, right?
- `read_all` doesn't return duplicates, right?
-  Only `read_many` could and that's because of line 39 `ids` argument that may contain duplicate, right?
- So if we really want to prevent that (not really sure if it makes a significant difference at this point :smile: ), we could `set(ids)` before passing them to `read_many`.

Reply:
I think it was not fully understood, maybe I need to put more comments:

- In the suggested code (with the `set(ids)` inside the for loop, you would be casting to set on *every* iteration, which in the long run might be costly.
- The idea behind having the `seen` set was not about duplicates (which cannot happen, ever unless something doggy happens). The point was to have some sort of "optimistic" optimization where there is no need to loop over the whooooole vocbulary, and stop when all ids have been seen.

Imagine the following case:

```
vocab.hits = [1, 2, 3, 4, 5 .... 1000000, 1000001]
ids = [1, 2, 3]
```

With the code as it was, it would stop and return in the 3rd iteration. With the suggested one it would loop 1000001 times. Obviously if the used asks for id 1000001 there is nothing we can do...

Reply:
Discuss IRL, leaving as is for now, we can revist in the future.